lightning/lightning.go

```diff
@@ -32,67 +34,105 @@ import (
 )
 
 type Lightning struct {
-	cfg      *config.Config
-	ctx      context.Context
-	shutdown context.CancelFunc
+	globalCfg *config.GlobalConfig
+	taskCfgs  chan *config.Config
+	ctx       context.Context
+	shutdown  context.CancelFunc
+	server    http.Server
+}
 
-	wg sync.WaitGroup
+func initEnv(cfg *config.GlobalConfig) error {
+	return log.InitLogger(&cfg.App.Config, cfg.TiDB.LogLevel)
 }
 
-func initEnv(cfg *config.Config) error {
-	if err := log.InitLogger(&cfg.App.Config, cfg.TiDB.LogLevel); err != nil {
-		return errors.Trace(err)
+func New(globalCfg *config.GlobalConfig) *Lightning {
+	if err := initEnv(globalCfg); err != nil {
+		fmt.Println("Failed to initialize environment:", err)
+		os.Exit(1)
 	}
 
-	if cfg.App.ProfilePort > 0 {
-		go func() {
-			http.Handle("/metrics", promhttp.Handler())
-			err := http.ListenAndServe(fmt.Sprintf(":%d", cfg.App.ProfilePort), nil)
-			log.L().Info("stopped HTTP server", log.ShortError(err))
-		}()
-	}
+	runtime.GOMAXPROCS(runtime.NumCPU())
 
-	return nil
+	ctx, shutdown := context.WithCancel(context.Background())
+	return &Lightning{
+		globalCfg: globalCfg,
+		ctx:       ctx,
+		shutdown:  shutdown,
+	}
 }
 
-func New(cfg *config.Config) *Lightning {
-	initEnv(cfg)
+func (l *Lightning) Serve() {
+	if len(l.globalCfg.App.StatusAddr) == 0 {
+		return
+	}
 
-	ctx, shutdown := context.WithCancel(context.Background())
+	http.Handle("/metrics", promhttp.Handler())
+	http.HandleFunc("/tasks", func(w http.ResponseWriter, req *http.Request) {
+		l.handleNewTask(w, req)
+	})
 
-	return &Lightning{
-		cfg:      cfg,
-		ctx:      ctx,
-		shutdown: shutdown,
+	l.server.Addr = l.globalCfg.App.StatusAddr
+	err := l.server.ListenAndServe()
+	log.L().Info("stopped HTTP server", log.ShortError(err))
+}
+
+// Run Lightning using the global config as the same as the task config.
+func (l *Lightning) RunOnce() error {
+	cfg := config.NewConfig()
+	if err := cfg.LoadFromGlobal(l.globalCfg); err != nil {
+		return err
 	}
+	if err := cfg.Adjust(); err != nil {
+		return err
+	}
+	return l.run(cfg)
 }
 
-func (l *Lightning) Run() error {
-	runtime.GOMAXPROCS(runtime.NumCPU())
-	common.PrintInfo("lightning", func() {
-		log.L().Info("cfg", zap.Stringer("cfg", l.cfg))
+func (l *Lightning) RunServer() error {
+	var finalErr error
+	l.taskCfgs = make(chan *config.Config, 64)
+
+	log.L().Info("Lightning server is running, post to /tasks to start an import task")
+
+	for {
+		select {
+		case <-l.ctx.Done():
+			return l.ctx.Err()
+		case task := <-l.taskCfgs:
+			if task == nil {
+				return finalErr
+			}
+			err := l.run(task)
+			if err != nil {
+				finalErr = err
+			}
+		}
+	}
+}
+
+var taskCfgRecorderKey struct{}
+
+func (l *Lightning) run(taskCfg *config.Config) error {
+	failpoint.Inject("SkipRunTask", func() {
```

Comment:
We can change the failpoint clousure signature to remove the `failpoint.Return` call.
```go
	failpoint.Inject("SkipRunTask", func() error {
		if recorder, ok := l.ctx.Value(&taskCfgRecorderKey).(chan *config.Config); ok {
			recorder <- taskCfg
		}
		return nil
	})
```